pubsub/azurepubsub/azurepubsub.go

```diff
@@ -0,0 +1,438 @@
+	}
+
+	subEntity, err = sm.Put(ctx, name, opts...)
+	if err != nil {		
+		return nil, err
+	}
+
+	return subEntity, nil
+}
+
+func deleteSubscription(ctx context.Context, topicName string, subscriptionName string, connectionString string) (error){
+	ns, err := getSbNamespace(connectionString)
+	if err != nil {
+		return err
+	}
+	sm, err := ns.NewSubscriptionManager(topicName)
+	se, err := sm.Get(ctx, subscriptionName)
+	if se != nil {
+		_ = sm.Delete(ctx, subscriptionName)
+	}
+	return nil
+}
+
+func (s *subscription) getSbSubscription(ctx context.Context)(*servicebus.Subscription, error){
+	sm, err := s.ns.NewSubscriptionManager(s.topicName)	
+	if err != nil {
+		return nil, err
+	}
+	// An empty SubscriptionEntity means no subscription exist for the given name. 
+	se, err := sm.Get(ctx, s.name)
+	if se == nil {
+		return nil, fmt.Errorf("azurepubsub: no such subscription %v", s.name)
+	}
+
+	sbTopic, err := s.ns.NewTopic(s.topicName)	
+	if err != nil {
+		return nil, err
+	}
+	return sbTopic.NewSubscription(s.name)
+}
+
+// IsRetryable implements driver.Subscription.IsRetryable.
+func (s *subscription) IsRetryable(error) bool {	
+	return false
+}
+
+// As implements driver.Subscription.As.
+func (s *subscription) As(i interface{}) bool {
+	p, ok := i.(**servicebus.Subscription)
+	if !ok {
+		return false
+	}	
+	*p, _ = s.getSbSubscription(context.Background())
+	return true
+}
+
+
+func (s *subscription) ErrorCode(err error) gcerrors.ErrorCode {
+	return errorCode(err)
+}
+
+// ReceiveBatch implements driver.Subscription.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {			
+	sbSub, err := s.getSbSubscription(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func(){
+		sbSub.Close(ctx)
+	}()
+	
+	receiverCtx, cancelCtx := context.WithTimeout(ctx, s.opts.ListenerTimeout)
+	defer cancelCtx()
+	
+	subReceiver, err := sbSub.NewReceiver(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func(){
+		subReceiver.Close(ctx)
+	}()
+	
+	var messages []*driver.Message			
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		subReceiver.Listen(receiverCtx, servicebus.HandlerFunc(func(iCtx context.Context, sbmsg *servicebus.Message)(error){		
+			metadata := map[string]string{}
+			for k, v := range sbmsg.UserProperties {			
+				metadata[k] = v.(string)
+			}			
+			messages = append(messages, &driver.Message{
+				Body:     sbmsg.Data,
+				Metadata: metadata,
+				AckID:    sbmsg.LockToken,
+			})			
+			if len(messages) >= maxMessages {
+				cancelCtx()
+			}
+			return nil
+		}))
+		
+		select {
+			case <- receiverCtx.Done():
+				wg.Done()
+		}
+	}() 
+
+	wg.Wait()		
+	return messages, nil
+}
+
+// SendAcks implements driver.Subscription.SendAcks.
+// IMPORTANT: This is a workaround to issue 'completed' message dispositions in bulk which is not supported in the ServiceBus SDK.  
+func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
```

Comment:
I think you want to put the receiver into PeekLockMode, then call Message.Complete here.

Reply:
The message is in PeakLockMode when it is Received (subscription.Receive(ctx, servicebus.HandlerFunc) where the LockDuration is based on the SB Subscription settings. SendAcks will issue the Complete disposition in bulk on a message that has already been Locked. Unlike the other providers, the message dispositions (complete, reject, deadlettered) are done at the message level. This is a workaround since the SB SDK does not allow re-hydrate of messages and bulk dispositions.

Reply:
Also as noted; the act of simply reading a message sets it in PeekLockMode, the other option is ReceiveAndDeleteMode which we don't want. If you read a message 10 times (can be configured) without issuing a Complete disposition, then it will be deadlettered.

Reply:
I guess I'm confused by why you're doing all this work in this method, including dialing a new connection, when as far as I can tell, all you need is to call Message.Complete for each message.

Maybe the key sentence is 

> This is a workaround since the SB SDK does not allow re-hydrate of messages and bulk dispositions.

but I don't understand what it means. Can you explain it in another way?

Reply:
I think I'm starting to understand. By "re-hydrate" do you mean "create a new instance of"? I was confused by that term.

It looks like you're saying that in order to call Complete, you need the same `*servicebus.Message` value that you got in the `Listen` callback. You can't just take the ack ID, make a new `*servicebus.Message` and call `Complete` on it, because it's missing the `ec` field. Is that right?

If so, then what if you set the AckID to the message? In other words, above where you now have
```
AckID:    sbmsg.LockToken
```
Change it to
```
AckID:    sbmsg
```
Then I think `SendAcks` could do 
```
ackID.(*servicebus.Message).Complete()
```
It might seem weird to put the whole message in the AckID field. But I don't think we (the pubsub package itself) really cares what's in that field. It's just a way to communicate information between ReceiveBatch and SendAcks.

Reply:
No dice, the message connection is tied to its receiver connection. The receiver connection is tied to the [context](https://github.com/google/go-cloud/pull/1207/files#diff-1baff8573393260a211a7549e41fce64R242). The connection resets on any subsequent Receive call. This also has implications on message.As which means users cant call any disposition operations outside a listener scope.

link detached, reason: *Error{Condition: amqp:link:detach-forced, Description: The link 'G15:4257294:54mIB6ptaFBSMVzUJEMh51R66JNIEq0uR1_FbGbQQthzmJYkybiwOw' is force detached by the broker due to errors occurred in consumer(link1075391). Detach origin: InnerMessageReceiver was closed.

Reply:
It seems that way. I did some additional testing, all calls to SB SDKs Subscription.Receive gets a new receiver (new connection).

Reply:
That is a problem, but maybe one we can leave for later. 

I'm not aware of any other outstanding problems, so I'll approve. Although the button will be green, our convention is to wait for all assignees to approve before submitting.

Reply:
I'll work on a subsequent PR for examples. What happened to the benchmark tests? I benchmarked this earlier; would like to see how it stacks up against the others.